### DISCUSSION TITLE
Dynamic Index and Milvus Questions

### DISCUSSION THREAD

**Discussion by Zijie-Tian** [milvus-io/milvus#26602] [Q&A and General discussion] [answered]
I've only recently been introduced to Milvus-related content. After reading Milvus's paper and some of its documentation, I've developed some confusion and would like to ask the following:

* Milvus seems to utilize LSM-Tree to handle insert and delete operations. Wouldn't this approach introduce significant latency during searches? How do you address such issues?
* I came across a statement saying "DiskANN is the default indexing method for Milvus." What does this mean?
* Does the DiskANN index also employ the LSM-Tree approach to manage insert and delete operations?

**Comment by yhmo** [comment, marked as answer]
Actually, insert/delete requests are processed in this way:
1. the client sends insert/delete requests, received by the Proxy node
2. The proxy node passes the insert/delete requests to the Pulsar service
3. Query node and data node subscribe the insert/delete requests from the Pulsar service, and accumulate the data in a buffer. If the user calls search(), it can search the vectors in this buffer.
4. Once the buffer size hits a threshold, or the user calls flush(), the buffer is flushed into storage
5. Each flush action will generate a new segment in storage. A background thread detects the new segment and merges it with other small segments.
6. Index node detects the new segments, and builds index for them(each segment has an independent index).
7. Once a segment is indexed, the query node will load its index to replace the original vectors buffer.

The basic rule is: if a segment has no index, load the original vectors in memory, search the vectors by brute-force; once a segment is indexed, load its index into memory to replace the original vectors, search its index.

The pulsar could be the bottleneck, or data node could be the bottleneck. If the data node is too busy to handle so many requests, the proxy node will refuse to accept any insert/delete request, until the data node catches up the progress.


I didn't see "DiskANN is the default indexing method for Milvus" in the official doc site, where did you see it?


DISKANN is a graph-based index for vectors. Also follow the basic rule: use index to search if a segment has index, else brute-force search on original vectors.

**Reply by Zijie-Tian** [reply]
Is the workload of the Query Node too heavy? do top-K on every segment?
Have you ever considered how to address the issue of excessive search load on the Query Node caused by too much streaming data?

**Reply by yhmo** [reply]
To make it a distributed system, we must split the data into segments so that it can be scalable. Since the data is in segments, you don't know the real topk items in which segment. So, we get topk from each segment and merge into a final topk.

There is always a contradiction between segment data and streaming data. Lots of factors need to be considered. Most of the time, current design is a trade-off between the factors.

**Reply by Zijie-Tian** [reply]
I see. The multi-shard method provides better parallelism and is easier to implement in distributed system. Isn't it meanless to build a full index, rather than multi-shard method?

**Reply by Zijie-Tian** [reply]
I mean, in practical vector database system.

**Comment by Zijie-Tian** [comment]
I checked and what I should see is ‘DiskANN is the default for on-disk Index.’, right? As shown in the picture below.
![4977636b6d10212bbfa6b18f34e9f6f](https://github.com/milvus-io/milvus/assets/22677086/418f9067-728c-4ba1-a978-f9b1babc9727)
Additionally, I have some questions. 
* Given Milvus's approach to handling insertions and deletions, wouldn't the search process require traversing through all the Segments? Could this become a bottleneck?
* Does current Milvus handle a lot of dynamic data? According to Milvus's processing method, how can the SLA be ensured? Because blocking requests would lead to significant tail latency, right?

**Reply by xiaofan-luan** [reply]
DiskANN requires very high performance disk (more than 100k qps is recommended)
Witht the high qps support, diskANN try to leverage perf by aio, but yes you are right, the expect latency of diskANN is far more than in memory index
